src/sparse_int_set.jl

```diff
@@ -0,0 +1,292 @@
+end
+
+length(s::SparseIntSet) = length(s.packed)
+
+# This makes sure that when adding (pushing) an Int,
+# it's respective page is allocated and put at the index of the reverse such that
+# pageid_offset works as intended.
+# Pages will be allocated only once, when pushing an Int that belongs to them.
+# Other not used pages (created during resize!) will be undefs until one Int belonging to them gets added. 
+function assure!(s::SparseIntSet, pageid)
+    if pageid > length(s.reverse)
+        diff = pageid - length(s.reverse)
+
+        resize!(s.reverse, pageid - 1)
+        p = zeros(Int, INT_PER_PAGE)
+        push!(s.reverse, p)
+        for i = 1:diff
+            push!(s.counters, 0)
+        end
+        return p, true
+    elseif !isassigned(s.reverse, pageid)
+        p = zeros(Int, INT_PER_PAGE)
+        @inbounds s.reverse[pageid] = p 
+        @inbounds s.counters[pageid] = 0 
+        return p, true
+    end
+    return @inbounds s.reverse[pageid], false
+end
+
+function push!(s::SparseIntSet, i::Integer)
+    i <= 0 && throw(DomainError("Only positive Ints allowed."))
+    pageid, offset = pageid_offset(s, i)
+    page, newly_created = assure!(s, pageid)
+    if newly_created || page[offset] == 0
+        @inbounds page[offset] = length(s) + 1
+        @inbounds s.counters[pageid] += 1
+        push!(s.packed, i)
+        return s
+    end
+    return s
+end
+push!(s::SparseIntSet, is::Integer...) = (for i in is; push!(s, i); end; return s)
+
+function dirty_pop!(s::SparseIntSet)
+    if isempty(s)
+        throw(ArgumentError("Cannot pop an empty set."))
+    end
+    id = pop!(s.packed)
+    pageid, offset = pageid_offset(s, id)
+    @inbounds s.reverse[pageid][offset] = 0
+    @inbounds s.counters[pageid] -= 1
+    return id
+end
+
+@propagate_inbounds function pop!(s::SparseIntSet)
+    id = dirty_pop!(s::SparseIntSet)
+    cleanup!(s)
+    return id
+end
+
+function dirty_pop!(s::SparseIntSet, id::Integer)
+    id < 0 && throw(ArgumentError("Int to pop needs to be positive."))
+
+    @boundscheck if !in(id, s)
+        throw(BoundsError(s, id))
+    end
+    @inbounds begin
+        packed_endid           = s.packed[end] 
+        from_page, from_offset = pageid_offset(s, id)
+        to_page, to_offset     = pageid_offset(s, packed_endid)
+
+        packed_id                         = s.reverse[from_page][from_offset]
+        s.packed[packed_id]               = packed_endid
+        s.reverse[to_page][to_offset]     = s.reverse[from_page][from_offset]
+        s.reverse[from_page][from_offset] = 0
+        s.counters[from_page] -= 1
+        pop!(s.packed)
+    end
+    return id
+end
+
+@propagate_inbounds function pop!(s::SparseIntSet, id::Integer)
+    id = dirty_pop!(s, id)
+    cleanup!(s)
+    return id
+end
+
+function pop!(s::SparseIntSet, id::Integer, default)
+    id < 0 && throw(ArgumentError("Int to pop needs to be positive."))
+    in(id, s) ? (@inbounds pop!(s, id)) : default
```

Comment:
For code clarity it is good to always state the return.
```suggestion
    return in(id, s) ? (@inbounds pop!(s, id)) : default
```